### ISSUE TITLE
Batch submission of notebooks

### ISSUE THREAD

**Issue by gzt5142** [hytest-org/hytest#100]
Thread into which discussion can flow regarding how to run notebooks 'headless'.

This is being investigated as a component of the jupyterbook build system.  If JB is to be used to construct the 'curated' user experience we looked at a while back, then JB needs to run on the ESIP/QHUB platform -- it will need to be able to access S3 buckets, start and run clusters, etc.  This is the only way that JB knows how to build the output: just run the notebook. 

JB will handle running the notebook and collecting outputs; in an ideal world, JB would do that each time we push a new release to the `main` branch of the hytest repo. That would make book-building automatic.  The default runners on GitHub are not meant for compute-intensive tasks, and won't be able to do the heavy lifting that our tutorial notebooks do.  The idea here is to have that github action run JB on our hub resources.

**Comment by gzt5142** [comment]
Option 1: `kbatch`

This is a way to batch-submit to a kubernetes cluster.  This looks promising, but there are a few hurdles....

First, the good news... 

Visit <https://jupyter.qhub.esipfed.org/hub/token> to establish an access token.  This will be the secret code to inform cluster security that your request is legit.  Make note of the token. From within a QHub terminal: 
```
conda activate pangeo
kbatch configure --kbatch-url="http://kbatch-kbatch-proxy.dev.svc.cluster.local" --token="bunchanumbers"
```
This will set up configuration such that future calls to `kbatch` made from within this terminal will connect to the right endpoint (specific to your cluster), with the security token you specify.  Submitting via `kbatch` from another host is another matter. We need the URL endpoint that the remote host (like a github action) will use to submit to our cluster.... and the cluster machine must be running. 

To submit a job: 
```
kbatch job submit \
  --name="Job-name-goes-here" \
  --image="ghcr.io/username/my-image:latest" \
  --command="['command', 'arg1', 'arg1', 'etc']" 
```
The `image` is a docker container that has the hytest conda environment all set up. I presume that such a container exists on ESIP/QHUB, but I am not able to locate its name/URL. 


To get `kbatch` to work: 

- Must know the URL of the kbatch endpoint from **off-cluster** invocation. 
- Must know the URL to the container into which the command will be injected and run.

**Comment by gzt5142** [comment]
Option 2: `kbatch cron job`

`kbatch` has the facility to set up jobs on a schedule using (effectively) `cron`.  To adopt this strategy, the model/architecture would have to shift a little. 

- Instead of spawning a JB build from a github action, it would be launched via `cron`.  The build would happen on a schedule, whether notebooks changed or not. 
- The kbatch endpoint is known (since the job is submitted/scheduled from a QHub terminal. 
- The container URL still must be known. 

With this sort of set-up, we have fewer moving parts and less (i.e. zero) coordination between github and the ESIP/QHUB platform.

**Comment by gzt5142** [comment]
Note that for both of the `kbatch` options mentioned above, the first step in the job must be to clone the repo.  The spawned clusters will not have access to the ESIP/QHUB filesystem... the repo has to be downloaded/cloned from GitHub fresh at each build.

This is not a problem (all gh-actions do this), but is useful to note that a submitted job will not account for any local changes -- only what's committed in the github `main` (or whatever branch we want) will be available.

**Comment by peterm790** [comment]
@rsignell-usgs yes the environment in the notebook must be set to a plan `ipykernel` and not `users:pangeo` etc... 

another thing that caught me out was that the link to the image `"ghcr.io/username/my-image:latest"` must not contain the protocol, ie don't include `https` at the beginning.

**Comment by rsignell-usgs** [comment]
Thanks @peterm790 !  I had totally forgotten about that 2nd item

**Comment by rsignell-usgs** [comment]
Can you give an example of a script that worked?

**Comment by gzt5142** [comment]
> Something about getting rid of the environment stored in the notebook metadata (e.g. users-pangeo)? Was that it?

I hit this issue (perhaps it is merely similar, and not the same issue) with `papermill`... I haven't got to the point of hitting it with `kbatch`.  I found that I needed to specify the kernel explicitly for  the `papermill` command to run the notebook correctly.. 

```
papermill -k python3 Notebook.ipynb Output.ipynb
```

but that was issued from the command line after activating `pangeo` environment.  Moving that into a batch submission is where I'm getting outfoxed. The main hurdle at moment is that I am not finding the container image I should be using. 

If there is a functioning model that I could look at, it would be very helpful. Thanks so much.

**Comment by peterm790** [comment]
Below is an example of a .yaml I used, the image linked is a replica of the `users:pangeo` environment. That being said I never quite got this to run reliably, something to do with the docker image not being pulled successfully.....

      name: "HRRR"
      image: "ghcr.io/peterm790/pangeo:latest"
      command:
          -"papermill"
          -"cronjob.ipynb"
      code: "cronjob.ipynb"

**Comment by gzt5142** [comment]
Thanks.. that helps a lot. 

It looks like you built and registered your own container.  That's helpful to know. . . 

And I believe you are issuing your `kbatch` command from a terminal on esip/qhub.  To set up the cron execution.  If that's not right, then I think I might need to look harder at the correct kbatch endpoint to use when submitting jobs from off-cluster.

**Comment by rsignell-usgs** [comment]
Note that if your kbatch job wants to write to S3, it needs credentials, which can be passed as env vars thusly:
``` 
08:59 $ more shared/users/notebooks/HRRR/hrrr_best.yaml
name: hrrr-awesome
image: ghcr.io/iameskild/ogc-demo:latest
command: 
  - papermill
  - hrrr_best_kbatch.ipynb
code: hrrr_best_kbatch.ipynb
env: 
  AWS_ACCESS_KEY_ID: "AKIAxxxxxxxx"
  AWS_SECRET_ACCESS_KEY: "9YIXxxxxxxxxxx"
  ```

**Comment by gzt5142** [comment]
re: submission from github actions via kbatch. 

The github runner must have kbatch installed, and it will need to be configured at run-time with the kbatch service URL. When I'm in my terminal on esip-qhub, the `kbatch-url` is an alias which resolves to the server I've started.   For an off-cluster host to submit, there needs to be a service URL which accepts jobs (authenticated via token). I'm guessing that would be something like `https://jupyter.qhub.esipfed.org/hub/services/kbatch/`.  My strong suspicion is that this service is not configured on the hub server. 

One other option for this workflow is to configure a cluster and register it as a 3rd party runner. Authentication and other controls would be in place (standard)... but this would cause github to send the job to our runner for execution rather than one of their runners. That sort of architecture would probably be cleaner -- but we need to undertake the job of building a runner with cluster resources, available for execution from a github event trigger. 

Argo can (likely) handle the event listener and trigger. I'm still learning how all of these pieces work -- but it seems that withe the right cluster parameters, we might be able to put Argo in charge of the jupterbook build when a github event fires.

**Comment by gzt5142** [comment]
The starting point for figuring out our own runners: 

<https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners>

**Comment by gzt5142** [comment]
> I'm guessing that would be something like `https://jupyter.qhub.esipfed.org/hub/services/kbatch/`. My strong suspicion is that this service is not configured on the hub server.

Based on today's thread in the esip-qhub slack, it seems like that is correct -- the service is not exposed for off-cluster submissions.

**Comment by gzt5142** [comment]
Broader question about architecture -- Is there a use case beyond github actions?  If so, then kbatch submissions from outside the cluster using an exposed endpoint would be super useful.

If the goal is to do cluster-specific things as a part of github (or gitlab) actions, then a clustered runner is likely the cleaner solution.  That has other implications, I'm sure... the authentication scheme being the most important.

**Comment by gzt5142** [comment]
Success in submitting `kbatch` jobs from within an esip terminal : 

The job YML: 
```yaml
name: "kbatchtest"
image: "ghcr.io/peterm790/pangeo:latest"
command:
  - "papermill"
  - "-k"
  - "python3"
  - "NWIS_dtypes.ipynb"
code: "./NWIS_dtypes.ipynb"
```

The job runs on a 'headless' cluster node, meaning that it doesn't have access to the file system.  SO... the output from that notebook was not captured (more accurately -- it was written to STDOUT, which means it went into the batch void). This is expected behavior for `kbatch` jobs. . . 

But it does have implications for a jupyterbook build.... JB build will, in many ways, be easier than trying to run notebooks in `papermill`.  JB handles all that and caches results using its own mechanism. But that cache needs a space on a file system. . . So we need to figure out how to get a filesystem underneath the process.  AND... we need to have that batch process clone the repo in order to get all of the files it needs to build the book -- to a 'local' file system accessible within the `kbatch` execution container.
 
So the submitted batch job should be either python or shell script to: 
- [ ] Mount a file system
- [ ] Clone the repo
- [ ] run jupyterbook build
- [ ] push built HTML files to publication platform (gh pages?)

**Comment by gzt5142** [comment]
@rsignell-usgs wrote: 
> Note that if your kbatch job wants to write to S3, it needs credentials,

This is an extra wrinkle that we need to figure out... While credentials can be passed to the kbatch job as described, the notebooks complicate matters.  So far this is a thought experiment; I've not proven this yet... but here's the problem as I understand it: 

- A job submitted via kbatch runs in a custom Docker image
- That docker image has a very limited context, specified at build-time
- The context does not include a user's home directory
- Many of the notebooks that we've built demonstrate how a reader might use AWS/S3, so those tutorial notebooks intentionally consult the AWS config and credentials stored in the user's home directory. 
- See <https://github.com/hytest-org/hytest/blob/main/model_evaluation/tutorials/nwm_streamflow/01_Data_Prep.ipynb> for an example of a tutorial which uses a `ConfigParser` to obtain credentials from the user's home directory. 
- A notebook deploying a cluster config using `ebdpy` or a home-grown equivalent to parse the `~/.aws/config` or `~/.aws/credentials` files will fail, since `$HOME` and `~/.aws/` will not exist within the container. 
- I suppose we **could** put credentials in the context of the Docker container, but that would put the credentials secrets into the open via a world-readable container image. 


I'm now wondering about the viability of batch-building the jupyter book. It seems like the two options are: 
1) put secrets in the published container
2) refactor the affected notebooks to preferentially use AWS credentials within the build environment.

There may be other options that I'm not thinking of. Suggestions welcome.

If the goal of the notebook is to show users how to do things on their own, a custom work-around for this batch environment feels contrary to that purpose.  The security implications of including credentials in the container context make that a non-starter, I would think.

**Comment by gzt5142** [comment]
Summary of conversation with @jesse-ross --- 

The plan moving forward: 
- We're going to build a hytest-curated docker image and register it with ghcr.io as hytest-org owned.
- Using that image, will build rudimentary jupyterbook (without cluster or S3 elements) via `kbatch` submission from the esip/qhub shell. 
- Determine an architecture/model for keeping track of the secrets needed for these notebooks to write their outputs to S3.  Those credentials are possibly different for different notebooks.
- Add cluster and/or s3-specific notebooks to the jupyterbook build  
- Pending the exposure of a kbatch endpoint from esip/qhub, we can try to submit the build job from remote location. 
- If successful, build a github action to auto-submit whenever notebook sources change.

**Comment by peterm790** [comment]
Hi @gzt5142,

I used this repository to build the docker image: https://github.com/iameskild/repo2registry 

This diff between that repo and my fork: https://github.com/iameskild/repo2registry/compare/main...peterm790:repo2registry:main , Highlights the changes that need to be made. You will need to set the github secrets (secrets.GHCR_USERNAME and secrets.GHCR_TOKEN) in your fork to point to your own docker username and password. Once that is done you can just upload a conda environment.yml file and run the already set up github workflow which will then generate the docker image and make it available as a github package.

**Comment by gzt5142** [comment]
After a chat with @jesse-ross , it seems like the correct way to do this is the repo2registery method directly from the repo.

**Comment by gzt5142** [comment]
All this docker-talk seems detailed and specific enough that I am moving it as a sub-task to its own issue: #138 . 

Once we have the docker sorted, can come back here to talk kbatch and jb build.

**Comment by gzt5142** [comment]
There is a complication here related to S3 that we need to figure out... 

As @rsignell-usgs pointed out here <https://github.com/hytest-org/hytest/issues/100#issuecomment-1300473324> AWS credentials can be established for the submitted job.  This will be necessary because the job submitted via kbatch will not be attached to a file system (so it can't read the credentials file). 

Fine..... but the submitted notebooks to run may have their own code to parse AWS credentials as part of the notebook tutorial. See https://github.com/hytest-org/hytest/blob/main/model_evaluation/tutorials/nwm_streamflow/01_Data_Prep.ipynb for an example... The notebook uses the `configparser` library to extract credentials from `$HOME/.aws/credentials`. 

Several notebooks demonstrate how to set AWS credentials in this way (and others).  Those will all fail as kbatch jobs, because no file system. But they will work fine if executed from a named user account on esip/qhub (when the filesystem is available). 

And so... we have a couple of options.  Probably more, but I can only think of two at the moment: 
* Manually build notebooks as a named user -- no kbatch needed.  This will be a manual process spawned by a real user. Notebooks will be able to find credentials using the config file.  This user must have the correct credentials for each notebook's access needs. 
* Build in a filesystem mounting operation (and installing a credentials file) as part of the kbatch job. This would give the running notebooks a means of looking for credentials. 

@rsignell-usgs, @amsnyder  -- any thoughts on a preferred solution? Seems like a short-term fix would be to make it all manual and evolve toward the second option.

**Comment by gzt5142** [comment]
OK... so it seems like kbatch and headless submission is a future feature... so I will close this issue. 

Jupyterbook build will be a manual process, so all of the notebooks will be run with the AWS_PROFILE credentials of the user running the `jbook build` process.  Notebooks should be able to run without manual intervention, and outputs will be captured for the book. 

Because these tutorial notebooks are demos, it would be really helpful if they could operate on small subsets of the 'real' data.  Will cut down on execution time.  Example:  some of my notebooks model how to work with all of the NWIS gages.  I clip this list/dataframe to only 100 gages -- the process is identical for the whole notebook, but the execution time is much reduced.
